From d5a1e71b8ffaad0aa85d5b4e44b224723107299a Mon Sep 17 00:00:00 2001
From: Atneya Nair <atneya@google.com>
Date: Thu, 29 Aug 2024 20:53:41 -0700
Subject: [PATCH] Validate full attr chain for recording

We currently only re-validate and listen for changes in the top of the
attr chain.

Listen across the whole chain, and compute the appop state
from the whole chain.

Test: atest CtsMediaAudioPermissionTestCases
Bug: 325912429
Flag: EXEMPT security
Change-Id: I4af2bbb7ecd3c5450115400c7d5fac63056dc0b8
---
 .../audiopolicy/service/AudioRecordClient.cpp | 37 ++++++++++++++++++-
 1 file changed, 36 insertions(+), 1 deletion(-)

diff --git a/services/audiopolicy/service/AudioRecordClient.cpp b/services/audiopolicy/service/AudioRecordClient.cpp
index 79a745863f..9b07f681cd 100644
--- a/services/audiopolicy/service/AudioRecordClient.cpp
+++ b/services/audiopolicy/service/AudioRecordClient.cpp
@@ -59,6 +59,41 @@ int getTargetSdkForPackageName(std::string_view packageName) {
 bool doesPackageTargetAtLeastU(std::string_view packageName) {
     return getTargetSdkForPackageName(packageName) >= __ANDROID_API_U__;
 }
+
+class AttrSourceItr {
+  public:
+    using iterator_category = std::forward_iterator_tag;
+    using difference_type = std::ptrdiff_t;
+    using value_type = AttributionSourceState;
+    using pointer = const value_type*;
+    using reference = const value_type&;
+
+    AttrSourceItr() : mAttr(nullptr) {}
+
+    AttrSourceItr(const AttributionSourceState& attr) : mAttr(&attr) {}
+
+    reference operator*() const { return *mAttr; }
+    pointer operator->() const { return mAttr; }
+
+    AttrSourceItr& operator++() {
+        mAttr = !mAttr->next.empty() ? mAttr->next.data() : nullptr;
+        return *this;
+    }
+
+    AttrSourceItr operator++(int) {
+        AttrSourceItr tmp = *this;
+        ++(*this);
+        return tmp;
+    }
+
+    friend bool operator==(const AttrSourceItr& a, const AttrSourceItr& b) = default;
+
+    static AttrSourceItr end() { return AttrSourceItr{}; }
+private:
+    const AttributionSourceState * mAttr;
+};
+
+
 } // anonymous
 
 // static
@@ -171,7 +206,7 @@ void OpRecordAudioMonitor::checkOp(bool updateUidStates) {
     using permission::AttrSourceIter::cend;
 
     const auto check = [&](int32_t op) -> bool {
-        return std::all_of(cbegin(mAttributionSource), cend(), [&](const auto& x) {
+        return std::all_of(AttrSourceItr{mAttributionSource}, AttrSourceItr::end(), [&](const auto& x) {
                     return mAppOpsManager.checkOp(op, x.uid,
                                                   VALUE_OR_FATAL(aidl2legacy_string_view_String16(
                                                           x.packageName.value_or("")))) ==
-- 
2.34.1

